### PR TITLE
Clear pending data when stream is closed by server.

### PR DESCRIPTION
**Pull request by yang-g**
Also add a test.

### CODE REVIEW

**Comment by grpc-testing** [comment]
```
****************************************************************

libgrpc.so

     VM SIZE              FILE SIZE
 ++++++++++++++ GROWIN ++++++++++++++
  [ = ]       0 [None]     +40  +0.0%

  [ = ]       0 TOTAL      +40  +0.0%


****************************************************************

libgrpc++.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]



```

**Comment by grpc-testing** [comment]
```
[trickle] No significant performance differences
```

**Comment by grpc-testing** [comment]
```
[microbenchmarks] No significant performance differences
```

**Comment by markdroth** [comment]
Reviewed 11 of 11 files at r1.
Review status: all files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---

*[src/core/ext/transport/chttp2/transport/chttp2_transport.cc, line 1889 at r1](https://reviewable.io:443/reviews/grpc/grpc/14080#-L3EAli9JCNHm_3CPHL3:-L3EAli9JCNHm_3CPHL4:bq66l5p) ([raw file](https://github.com/grpc/grpc/blob/256a3705de5d9fd1eefad0ac08664e204ccd55f4/src/core/ext/transport/chttp2/transport/chttp2_transport.cc#L1889)):*
> ```cpp
>   if (s->recv_trailing_metadata_finished != nullptr && s->read_closed &&
>       s->write_closed) {
>     if (s->seen_error || t->is_client == false) {
> ```

Nit: Prefer `!` instead of `== false`.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/grpc/grpc/14080)*

**Comment by markdroth** [comment]
This looks reasonable to me, but I'd like Noah to review as well, since he's the new owner of the chttp2 code.

**Comment by yang-g** [comment]
ping @ncteisen : there is an internal user waiting for this fix :)

**Comment by grpc-testing** [comment]
```
****************************************************************

libgrpc.so

     VM SIZE              FILE SIZE
 ++++++++++++++ GROWIN ++++++++++++++
  [ = ]       0 [None]     +32  +0.0%

  [ = ]       0 TOTAL      +32  +0.0%


****************************************************************

libgrpc++.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]



```

File: test/cpp/end2end/server_early_return_test.cc [hunk 1]
```diff
@@ -0,0 +1,233 @@
...
+    }
+
+    gpr_sleep_until(gpr_time_add(
+        gpr_now(GPR_CLOCK_MONOTONIC),
+        gpr_time_from_micros(server_delay_before_return_us, GPR_TIMESPAN)));
+
+    return Status(static_cast<StatusCode>(server_return_status_code), "");
+  }
+
+  int GetIntValueFromMetadata(ServerContext* context, const char* key,
```
Comment (ncteisen) [inline test/cpp/end2end/server_early_return_test.cc]: This is copied from end2end and async_end2end right? Nit, and no need to do it in this PR, but we should create a helper lib with this and other util functions now that we have it in three places

**Comment by grpc-testing** [comment]
```
[trickle] No significant performance differences
```

**Comment by grpc-testing** [comment]
```
****************************************************************

libgrpc.so

     VM SIZE              FILE SIZE
 ++++++++++++++ GROWIN ++++++++++++++
  [ = ]       0 [None]     +32  +0.0%

  [ = ]       0 TOTAL      +32  +0.0%


****************************************************************

libgrpc++.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]



```

**Comment by grpc-testing** [comment]
```
[trickle] No significant performance differences
```

**Comment by grpc-testing** [comment]
```
[microbenchmarks] No significant performance differences
```

**Comment by grpc-testing** [comment]
```
[microbenchmarks] No significant performance differences
```

**Comment by yang-g** [comment]
Test failures: #13477

**Pull request merged**